pythran/pythonic/numpy/base_repr.hpp

```diff
@@ -42,11 +44,10 @@ namespace numpy
 
   types::str base_repr(long number, long base, long padding)
   {
-    char *mem = new char[sizeof(number) * 8 + 1 + padding];
-    std::fill(mem, mem + padding, '0');
-    details::itoa(number, mem + padding, base);
-    auto res = types::str(mem);
-    delete[] mem;
+    std::unique_ptr<char[]> mem{new char[sizeof(number) * 8 + 1 + padding]};
+    std::fill(mem.get(), mem.get() + padding, '0');
+    details::itoa(number, mem.get() + padding, base);
+    auto res = types::str(mem.get());
```

Comment:
looking at this, it would be nice to have a constructor of `types::str` capable of taking a `unique_ptr` and moving from it

Reply:
You even have a comment about that in pythran/pythonic/types/file.hpp:

```c++
// This part needs a new implementation of types::str(char*) to avoid
// unnecessary copy.
```

Maybe my next little task ;)

Reply:
Is far as I know, you cannot give an existing buffer to std::string (which you use as underlying container type).

What can be done is allocate a string, resize it and work on its internal buffer. This is not the most  pretty thing I have done, but it works.

It could look like:

```diff
diff --git a/pythran/pythonic/include/types/str.hpp b/pythran/pythonic/include/types/str.hpp
index 0849014da..ab23a2711 100644
--- a/pythran/pythonic/include/types/str.hpp
+++ b/pythran/pythonic/include/types/str.hpp
@@ -133,6 +133,7 @@ namespace types
 
     types::str &operator+=(types::str const &s);
 
+    container_type &get_data();
     container_type const &get_data() const;
 
     long size() const;
diff --git a/pythran/pythonic/numpy/base_repr.hpp b/pythran/pythonic/numpy/base_repr.hpp
index 07a44659f..4893c0d31 100644
--- a/pythran/pythonic/numpy/base_repr.hpp
+++ b/pythran/pythonic/numpy/base_repr.hpp
@@ -6,8 +6,6 @@
 #include "pythonic/utils/functor.hpp"
 #include "pythonic/types/ndarray.hpp"
 
-#include <memory>
-
 PYTHONIC_NS_BEGIN
 
 namespace numpy
@@ -44,10 +42,11 @@ namespace numpy
 
   types::str base_repr(long number, long base, long padding)
   {
-    std::unique_ptr<char[]> mem{new char[sizeof(number) * 8 + 1 + padding]};
-    std::fill(mem.get(), mem.get() + padding, '0');
-    details::itoa(number, mem.get() + padding, base);
-    auto res = types::str(mem.get());
+    types::str res;
+    res.resize(sizeof(number) * 8 + 1 + padding);
+    std::fill((char *)res.get_data().data(),
+              (char *)res.get_data().data() + padding, '0');
+    details::itoa(number, (char *)res.get_data().data() + padding, base);
     return res;
   }
 }
diff --git a/pythran/pythonic/types/file.hpp b/pythran/pythonic/types/file.hpp
index e4064a1ea..5f559ba9a 100644
--- a/pythran/pythonic/types/file.hpp
+++ b/pythran/pythonic/types/file.hpp
@@ -20,7 +20,6 @@
 #include <string>
 #include <cstdio>
 #include <unistd.h>
-#include <memory>
 
 PYTHONIC_NS_BEGIN
 
@@ -159,11 +158,10 @@ namespace types
     seek(0, SEEK_END);
     size = size < 0 ? tell() - curr_pos : size;
     seek(curr_pos);
-    std::unique_ptr<char[]> content{new char[size + 1]};
-    // This part needs a new implementation of types::str(char*) to avoid
-    // unnecessary copy.
-    types::str res(content.get(),
-                   fread(content.get(), sizeof(char), size, **data));
+    types::str res;
+    res.resize(size + 1);
+    res.resize(
+        fread((char *)res.get_data().data(), sizeof(char), size, **data));
     return res;
   }
 
diff --git a/pythran/pythonic/types/str.hpp b/pythran/pythonic/types/str.hpp
index a035976e5..97a072e00 100644
--- a/pythran/pythonic/types/str.hpp
+++ b/pythran/pythonic/types/str.hpp
@@ -357,6 +357,11 @@ namespace types
     return *this;
   }
 
+  str::container_type &str::get_data()
+  {
+    return *data;
+  }
+
   str::container_type const &str::get_data() const
   {
     return *data;
```

Would you want this approach to avoid completly allocating a intermediate buffer?

Reply:
I like the approach :-)

Reply:
After further check, this will only work with c++17 and later.

In earlier version of the language, non const version of `char* std::string::data()` does not exist. So to support this it would require a `const_cast`.